src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -9,6 +9,174 @@ namespace Microsoft.ML.Runtime.Internal.CpuMath
+        {
+            Contracts.Assert(mat.Size == dst.Size * src.Size);
+
+            if (Sse.IsSupported)
+            {
+                if (!tran)
+                {
+                    Contracts.Assert(0 <= crun && crun <= dst.Size);
+                    SseIntrinsics.MatMulA(add, mat, src, dst, crun, src.Size);
+                }
+                else
+                {
+                    Contracts.Assert(0 <= crun && crun <= src.Size);
+                    SseIntrinsics.MatMulTranA(add, mat, src, dst, dst.Size, crun);
+                }
+            }
+            else
+            {
+                if (!tran)
+                {
+                    Contracts.Assert(0 <= crun && crun <= dst.Size);
+                    for (int i = 0; i < crun; i++)
+                    {
+                        float dotProduct = 0;
+                        for (int j = 0; j < src.Size; j++)
+                        {
+                            dotProduct += mat[i * src.Size + j] * src[j];
+                        }
+
+                        if (add)
+                        {
+                            dst[i] += dotProduct;
+                        }
+                        else
+                        {
+                            dst[i] = dotProduct;
+                        }
+                    }
+                }
+                else
+                {
+                    Contracts.Assert(0 <= crun && crun <= src.Size);
+                    for (int i = 0; i < dst.Size; i++)
+                    {
+                        float dotProduct = 0;
+                        for (int j = 0; j < crun; j++)
+                        {
+                            dotProduct += mat[j * src.Size + i] * src[j];
+                        }
+
+                        if (add)
+                        {
+                            dst[i] += dotProduct;
+                        }
+                        else
+                        {
+                            dst[i] = dotProduct;
+                        }
+                    }
+                }
+            }
+        }
+
+        public static void MatTimesSrc(bool tran, bool add, AlignedArray mat, int[] rgposSrc, AlignedArray srcValues,
+            int posMin, int iposMin, int iposLim, AlignedArray dst, int crun)
+        {
+            Contracts.AssertValue(rgposSrc);
+            Contracts.Assert(0 <= iposMin && iposMin <= iposLim && iposLim <= rgposSrc.Length);
+            Contracts.Assert(mat.Size == dst.Size * srcValues.Size);
+
+            if (iposMin >= iposLim)
+            {
+                if (!add)
+                    dst.ZeroItems();
+                return;
+            }
+
+            Contracts.AssertNonEmpty(rgposSrc);
+
+            if (Sse.IsSupported)
+            {
+                if (!tran)
+                {
+                    Contracts.Assert(0 <= crun && crun <= dst.Size);
```

Comment:
Here also, and all the places where this can be unified and change the comparison to be `crun >= 0`